flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2016, Triad National Security, LLC
+// All rights reserved.
+
+#ifndef FLECSI_EXEC_HPX_TASK_PROLOGUE_HH
+#define FLECSI_EXEC_HPX_TASK_PROLOGUE_HH
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+#include <hpx/modules/concurrency.hpp>
+#include <hpx/modules/serialization.hpp>
+
+#include "flecsi/data/privilege.hh"
+#include "flecsi/data/topology.hh"
+#include "flecsi/exec/hpx/bind_accessors.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/demangle.hh"
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace flecsi {
+namespace topo {
+struct global_base;
+}
+
+namespace exec {
+
+struct task_prologue_base {
+
+protected:
+  // Those methods are "protected" because they are *only* called by
+  // flecsi::exec::prolog which inherits from task_prologue.
+
+  // Patch up the "un-initialized" type conversion from future<R, index> to
+  // future<R, single> in the generic code.
+  template<typename R>
+  static void visit(future<R, exec::launch_type_t::single> & single,
+    future<R, exec::launch_type_t::index> & index) {
+    single = index.get(flecsi::run::context::instance().color());
+  }
+
+  // visit generic topology
+  template<typename T,
+    Privileges P,
+    typename Topo,
+    typename Topo::index_space Space>
+  void visit(data::accessor<data::raw, T, P> &,
+    data::field_reference<T, data::raw, Topo, Space> const & ref) {
+
+    const field_id_t f = ref.fid();
+    constexpr bool glob =
+      std::is_same_v<typename Topo::base, topo::global_base>;
+
+    auto & t = ref.topology();
+    auto & r_or_p = [&]() -> auto & {
+      if constexpr(glob) {
+        return *t;
+      }
+      else {
+        // The partition controls how much memory is allocated.
+        return *t.template get_partition<Space>();
+      }
+    }
+    ();
+
+    // store associated partition for bind_accessors
+    regions_partitions.push_back(&r_or_p);
+
+    auto & field = r_or_p[f];
+    bool added_as_dependency = false;
+
+    // Note that this can, even if P is all read-only, reentrantly post a task
+    // that writes to the field (because it needs to update ghost values before
+    // the user task reads them).
+    data::region & reg = t.template get_region<Space>();
+    if constexpr(!glob) {
+      // Create a new task that performs the required ghost-copy and make the
+      // task currently being created depend on the results of the ghost-copy
+      // operation. This happens inside ghost_copy (see the implementation
+      // copy_engine::operator()()).
+      reg.ghost_copy<P>(ref);
+
+      // add any pending ghost_copy operations as a dependency for this task
+      if(field.dep == data::dependency::ghost_copy) {
```

Comment:
This value serves as a marker that the ghost-copy created a future that needs to be stored as a dependency for the current task. If the ghost-copy would just set the field's dependency type to read or write, we wouldn't be able to find out whether the field's future needs to be added to the dependency list.

Reply:
How is the future created by the ghost copy distinct from that which would be created by an ordinary task writing to the same field?  It would (in the long term) actually be useful for it to be different, in that a task which requested `na` for ghosts and `ro` for shared can run concurrently with the ghost copy, but I don't think the HPX backend implements that any more than the Legion one does at the moment.

Reply:
The futures themselves are indistinguishable.

Reply:
Sure, but why do they _need_ to be distinguished?

Reply:
If the field already had a future associated (before the ghost-copy), we don't need to add it to the dependencies of the current task. Only futures that are created for the ghost-copy itself should be added as a dependency here.

Reply:
If it had a future associated because it was being (say) written by another task, and there is no ghost copy at all, we very much have to add that future as a dependency, right?  Are you saying that you need to know whether the call to `ghost_copy` actually did something because it influences the dependencies in some way not captured by the dependency on any prior task?

Reply:
> If it had a future associated because it was being (say) written by another task, and there is no ghost copy at all, we very much have to add that future as a dependency, right? 

Yes, we do, but not at this spot. Here we worry about the ghost-copy only. 

> Are you saying that you need to know whether the call to `ghost_copy` actually did something because it influences the dependencies in some way not captured by the dependency on any prior task?

Exactly, yes.

Reply:
I thought I had read and understood the code, but apparently not.  If you'll pardon my inattentiveness, could you explain why it should (need to) make a difference?

Reply:
I might have been a bit overcautious here by assuming the ghost-copy somehow influences the dependency handling in special ways. You're right - it does not. 

There are obviously two scenarios: 

- the field has already a future associated with it from a previous (pending) operation.
- the field has no (pending) operations 

The former case is handled by using the future as a dependency for the ghost-copy operation (see [here](https://github.com/flecsi/flecsi/blob/aa78b61684df095924d2bdebe78351d91193ebeb/flecsi/data/hpx/copy.hh#L61)).
In the latter case, the ghost-copy operation creates a new future (see [here](https://github.com/flecsi/flecsi/blob/aa78b61684df095924d2bdebe78351d91193ebeb/flecsi/data/hpx/copy.hh#L70)).

I both cases we can use the returned future as a dependency for the new task. So, there is no need to distinguish the 'type' of the future. I will adapt the code accordingly.

Reply:
Right, but then the field will have whatever previous dependency stored there, on which we might still need to wait, right?

Reply:
Ok, but nevertheless, in this case we shouldn't change the dependency type of the pre-existing future.

Reply:
Right: and since `reg.ghost_copy()` doesn't do anything, no assignment to the dependency type will happen (in `hpx/copy.hh`).